### PR TITLE
ASoC: SOF: Intel: BYT: refine the IPC interrupt handling

### PR DESCRIPTION
**Pull request by keyonjie**
This is to fix https://github.com/thesofproject/linux/issues/1492

We should mask interrupt once we got it in the handler, otherwise, we
may get interrupt storm in case where the irq thread has no chance to be
scheduled.

Signed-off-by: Keyon Jie <yang.jie@linux.intel.com>

### CODE REVIEW

**Comment by keyonjie** [comment]
I have only tested this on BSW so wait to see the on device test result on baytrail from the CI.

File: sound/soc/sof/intel/byt.c [hunk 1]
```diff
@@ -203,16 +220,11 @@ static irqreturn_t byt_irq_thread(int irq, void *context)
...
-	if (ipcx & SHIM_BYT_IPCX_DONE &&
-	    !(imrx & SHIM_IMRX_DONE)) {
-		/* Mask Done interrupt before first */
-		snd_sof_dsp_update_bits64_unlocked(sdev, BYT_DSP_BAR,
-						   SHIM_IMRX,
-						   SHIM_IMRX_DONE,
-						   SHIM_IMRX_DONE);
-
+	if ((ipcx & SHIM_BYT_IPCX_DONE) &&
+	    (imrx & SHIM_IMRX_DONE)) {
```
Comment (plbossart) [inline sound/soc/sof/intel/byt.c:228]: why are you changing this?

File: sound/soc/sof/intel/byt.c [hunk 2]
```diff
@@ -231,15 +243,8 @@ static irqreturn_t byt_irq_thread(int irq, void *context)
...
-	if (ipcd & SHIM_BYT_IPCD_BUSY &&
-	    !(imrx & SHIM_IMRX_BUSY)) {
-		/* Mask Busy interrupt before return */
-		snd_sof_dsp_update_bits64_unlocked(sdev, BYT_DSP_BAR,
-						   SHIM_IMRX,
-						   SHIM_IMRX_BUSY,
-						   SHIM_IMRX_BUSY);
-
+	if ((ipcd & SHIM_BYT_IPCD_BUSY) &&
+	    (imrx & SHIM_IMRX_BUSY)) {
```
Comment (plbossart) [inline sound/soc/sof/intel/byt.c:248]: why are you changing this?

File: sound/soc/sof/intel/byt.c [hunk 3]
```diff
@@ -184,13 +184,30 @@ static void byt_dump(struct snd_sof_dev *sdev, u32 flags)
...
-	if (isr & (SHIM_ISRX_DONE | SHIM_ISRX_BUSY))
+	ipcx = snd_sof_dsp_read64(sdev, BYT_DSP_BAR, SHIM_IPCX);
+	ipcd = snd_sof_dsp_read64(sdev, BYT_DSP_BAR, SHIM_IPCD);
+
+	if ((isr & SHIM_ISRX_DONE) && (ipcx & SHIM_BYT_IPCX_DONE)) {
+		/* reply message from DSP, Mask Done interrupt first */
+		snd_sof_dsp_update_bits64_unlocked(sdev, BYT_DSP_BAR,
+						   SHIM_IMRX,
+						   SHIM_IMRX_DONE,
+						   SHIM_IMRX_DONE);
```
Comment (plbossart) [inline sound/soc/sof/intel/byt.c:200]: I am not touching this PR without a document explaining the ideas.

**Comment by keyonjie** [comment]
> I am not even going to look at this as is.
> You are changing the entire IPC without a shred of explanations and using registers in completely different ways.
> Please.
> Let's be serious, shall we?

Sorry, I foresaw this challenge, but I thought I had explained clearly in the commit message.

Let me think more about how to make it more straight-forward.

**Comment by ranj063** [comment]
@keyonjie why not disable IPC interrupt (IMRX/IMRD?) in the handler and disable in the thread just like we do for HDA?

**Comment by keyonjie** [comment]
> @keyonjie why not disable IPC interrupt (IMRX/IMRD?) in the handler and disable in the thread just like we do for HDA?

Yes and no here.

"why not disable IPC interrupt (IMRX/IMRD?) in the handler?"

Yes for this, my PR is changing this to align with SKL+.

"and disable in the thread"

No for this in my PR, In BYT, there isn't a common bit for enabling/disabling IPC interrupt like HDA_DSP_ADSPIS_IPC in HDA case, that means we have to use IMRX for that.

**Comment by ranj063** [comment]
> No for this in my PR, In BYT, there isn't a common bit for enabling/disabling IPC interrupt like HDA_DSP_ADSPIS_IPC in HDA case, that means we have to use IMRX for that.

And is that a problem?

**Comment by plbossart** [comment]
@keyonjie  We've worked with BYT for a very long time, so the bar to change anything is very very high now.
I don't see why we would re-interpret the meaning of registers.
If something is broken we fix it, but we don't reinvent the IPC.

**Comment by keyonjie** [comment]
Hi @plbossart @ranj063 I just filed an issue here:
https://github.com/thesofproject/linux/issues/1492

Where you will see the issue very straight-forward.

**Comment by keyonjie** [comment]
> @keyonjie We've worked with BYT for a very long time, so the bar to change anything is very very high now.
> I don't see why we would re-interpret the meaning of registers.
> If something is broken we fix it, but we don't reinvent the IPC.

Understood, sorry I worked on SKL+ at that time and didn't pay more attention to BYT refining.

Please see the issue I filed here: https://github.com/thesofproject/linux/issues/1492
It is obvious we have critical issues here, and this is why the system is hang in module unload/reload on BSW.

**Comment by plbossart** [comment]
> > @keyonjie We've worked with BYT for a very long time, so the bar to change anything is very very high now.
> > I don't see why we would re-interpret the meaning of registers.
> > If something is broken we fix it, but we don't reinvent the IPC.
> 
> Understood, sorry I worked on SKL+ at that time and didn't pay more attention to BYT refining.
> 
> Please see the issue I filed here: #1492
> It is obvious we have critical issues here, and this is why the system is hang in module unload/reload on BSW.

I've run the module load/unload tests on Baytrail before, no issues.
can you please retest on a regular baytrail device, e.g. Minnowboard, to make sure your assessment is indeed correct?

**Comment by keyonjie** [comment]
@plbossart as I commented in the issue #1492, both CHT and minnow board have the similar risk I described on #1492, we got multiple entries to the interrupt handler for a single interrupt, though it looks fine at module unload/reload on them.

That's why we need to change the IPC interrupt handling here.

see the dmesg logs below:

On CHT:

[    6.469040] sof-audio-acpi 808622A8:00: ipc tx: 0x30030000
[    6.469065] Keyon: byt_irq_handler, 192, isr:0x1
[    6.469091] Keyon: byt_irq_handler, 192, isr:0x1
[    6.469106] Keyon: byt_irq_thread, 210, imrx:0x0, ipcx:0x4000000000000000, ipcd:0x70028800
[    6.469113] Keyon: byt_irq_handler, 192, isr:0x1
[    6.469119] Keyon: byt_irq_thread, 216
[    6.469156] Keyon: byt_irq_handler, 192, isr:0x0
[    6.469171] Keyon: byt_irq_thread, 210, imrx:0x0, ipcx:0x0, ipcd:0x70028800
[    6.469175] sof-audio-acpi 808622A8:00: ipc tx succeeded: 0x30030000
[    6.469191] sof-audio-acpi 808622A8:00: sink PGA3.0 control none source BUF3.0
[    6.469202] sof-audio-acpi 808622A8:00: ipc tx: 0x30030000
[    6.469223] Keyon: byt_irq_handler, 192, isr:0x1
[    6.469252] Keyon: byt_irq_handler, 192, isr:0x1
[    6.469268] Keyon: byt_irq_thread, 210, imrx:0x0, ipcx:0x4000000000000000, ipcd:0x70028800
[    6.469276] Keyon: byt_irq_handler, 192, isr:0x1
[    6.469282] Keyon: byt_irq_thread, 216
[    6.469318] Keyon: byt_irq_handler, 192, isr:0x0
[    6.469335] sof-audio-acpi 808622A8:00: ipc tx succeeded: 0x30030000
[    6.469352] sof-audio-acpi 808622A8:00: sink BUF3.1 control none source PGA3.0
[    6.469355] Keyon: byt_irq_thread, 210, imrx:0x0, ipcx:0x0, ipcd:0x70028800
[    6.469364] sof-audio-acpi 808622A8:00: ipc tx: 0x30030000
[    6.469392] Keyon: byt_irq_handler, 192, isr:0x1
[    6.469420] Keyon: byt_irq_handler, 192, isr:0x1
[    6.469447] Keyon: byt_irq_handler, 192, isr:0x1
[    6.469461] Keyon: byt_irq_thread, 210, imrx:0x0, ipcx:0x4000000000000000, ipcd:0x70028800
[    6.469468] Keyon: byt_irq_handler, 192, isr:0x1
[    6.469474] Keyon: byt_irq_thread, 216
[    6.469485] Keyon: byt_irq_handler, 192, isr:0x1
[    6.469529] Keyon: byt_irq_thread, 210, imrx:0x0, ipcx:0x0, ipcd:0x70028800
On Minnow board,

[   12.833394] sof-audio-acpi 80860F28:01: booting DSP firmware
[   12.835286] Keyon: byt_irq_handler, 192, isr:0x2
[   12.835380] Keyon: byt_irq_handler, 192, isr:0x2
[   12.835415] Keyon: byt_irq_handler, 192, isr:0x2
[   12.835432] Keyon: byt_irq_handler, 192, isr:0x2
[   12.835449] Keyon: byt_irq_handler, 192, isr:0x2
[   12.835466] Keyon: byt_irq_handler, 192, isr:0x2
[   12.835484] Keyon: byt_irq_handler, 192, isr:0x2
[   12.835509] Keyon: byt_irq_handler, 192, isr:0x2
[   12.835532] Keyon: byt_irq_handler, 192, isr:0x2
[   12.835550] Keyon: byt_irq_handler, 192, isr:0x2
[   12.835559] Keyon: byt_irq_thread, 210, imrx:0x0, ipcx:0x4000, ipcd:0x8000000070028800
[   12.835567] Keyon: byt_irq_handler, 192, isr:0x2
[   12.835591] Keyon: byt_irq_thread, 243
[   12.835597] Keyon: byt_irq_handler, 192, isr:0x2
[   12.835619] sof-audio-acpi 80860F28:01: ipc rx: 0x70000000
[   12.835628] sof-audio-acpi 80860F28:01: ipc: DSP is ready 0x70000000 offset 0x144000
[   12.835657] sof-audio-acpi 80860F28:01: Firmware info: version 1:1:0-f004d
[   12.835665] sof-audio-acpi 80860F28:01: Firmware: ABI 3:11:0 Kernel ABI 3:11:0
[   12.835673] sof-audio-acpi 80860F28:01: Firmware debug build 1 on Nov 14 2019-21:16:33 - options:
```

**Comment by plbossart** [comment]
@keyonjie Can we first try and understand the initial blocker, which was to deal with module load/unload? There does not seem to be any correlation nor causation between IPC issues and module removals.

**Comment by keyonjie** [comment]
> @keyonjie Can we first try and understand the initial blocker, which was to deal with module load/unload? There does not seem to be any correlation nor causation between IPC issues and module removals.

Per my understanding, there could be, I believe the scenario we hang there is because of a IPC interrupt storm happened. The fact that applying these will fix the hang issue proved that.

And to say the least, I actually can't bear this kind of obvious risk in the code, can you?

**Comment by plbossart** [comment]
> > @keyonjie Can we first try and understand the initial blocker, which was to deal with module load/unload? There does not seem to be any correlation nor causation between IPC issues and module removals.
> 
> Per my understanding, there could be, I believe the scenario we hang there is because of a IPC interrupt storm happened. The fact that applying these will fix the hang issue proved that.
> 
> And to say the least, I actually can't bear this kind of obvious risk in the code, can you?

In God we trust, others bring data. 

Your theory of IPC rain storm may be right, but why doesn't it impact other platforms then? It needs to be backed by either experimental evidence or demonstration that the hang is removed with your fixes AND there is no regression on other platforms.

I am not going to change one line of IPC code without careful explanation and extensive testing. The code works, maybe not optimally, on all other platforms, and has done so for many months, so the 'obvious risk' is not so obvious, sorry. The last 'obvious' IPC fix was reverted due to other issues which exposed an **obvious** lack of extensive testing. Not going to happen twice, sorry. 

It's fine to experiment, it's a completely different story to submit a PR and ask that it be merged. This PR does not have a 'Draft' or 'TEST' or 'RFC' status, so I will push back if I don't feel comfortable with the its maturity.

**Comment by keyonjie** [comment]
> > > @keyonjie Can we first try and understand the initial blocker, which was to deal with module load/unload? There does not seem to be any correlation nor causation between IPC issues and module removals.
> > 
> > 
> > Per my understanding, there could be, I believe the scenario we hang there is because of a IPC interrupt storm happened. The fact that applying these will fix the hang issue proved that.
> > And to say the least, I actually can't bear this kind of obvious risk in the code, can you?
> 
> In God we trust, others bring data.
> 
> Your theory of IPC rain storm may be right, but why doesn't it impact other platforms then? It needs to be backed by either experimental evidence or demonstration that the hang is removed with your fixes AND there is no regression on other platforms.
> 
> I am not going to change one line of IPC code without careful explanation and extensive testing. The code works, maybe not optimally, on all other platforms, and has done so for many months, so the 'obvious risk' is not so obvious, sorry. The last 'obvious' IPC fix was reverted due to other issues which exposed an **obvious** lack of extensive testing. Not going to happen twice, sorry.
> 
> It's fine to experiment, it's a completely different story to submit a PR and ask that it be merged. This PR does not have a 'Draft' or 'TEST' or 'RFC' status, so I will push back if I don't feel comfortable with the its maturity.

I totally understand your point and I appreciate to that, IPC is a crucial part so any change to it should pass stress test, this is the right attitude to guarantee our code quality.

Let me change the subject and call for more test on BYT/CHT/BSW.

**Comment by keyonjie** [comment]
@mengdonglin @keqiaozhang Can I ask for a stress/extensive test on BTY/CHT/BSW with this PR applied, it is not expected to fix issues on BYT/CHT, so no regression with it applied is good enough to me.

**Comment by fredoh9** [comment]
I ran one hour of stress-test. I don't see any regression issue.

**Comment by keyonjie** [comment]
@fredoh9 Thank you. @plbossart @ranj063 The test on my side looks good on cht and bsw also.

**Comment by keyonjie** [comment]
@plbossart @ranj063 Can you consider taking this one please, the logs in #1492 shows the risk without this and test shows no regression with the PR.

**Comment by plbossart** [comment]
> @plbossart @ranj063 Can you consider taking this one please, the logs in #1492 shows the risk without this and test shows no regression with the PR.

no. I don't have a full picture of why this is needed and what your fix accomplishes and what it actually fixes, and if we have similar issues with other platforms.

**Comment by keyonjie** [comment]
> > @plbossart @ranj063 Can you consider taking this one please, the logs in #1492 shows the risk without this and test shows no regression with the PR.
> 
> no. I don't have a full picture of why this is needed and what your fix accomplishes and what it actually fixes, and if we have similar issues with other platforms.

Why this is needed:
With today's code, on BYT, byt_irq_handler() is entered multiple times for a single IPC interrupt, as we don't mask the interrupt in time in the handler, the next interrupt handler entry(caused by the same IPC interrupt) could happen before the previous interrupt thread mask the interrupt and preempt it. The test code and test result demonstrate this.
Not talking about the consequence and risk of it, this "byt_irq_handler() is entered multiple times for a single IPC interrupt" goes against our design original intention, isn't it?

What my fix accomplishes:
With moving the masking of the SHIM_IMRX_BUSY/DONE bits from the interrupt thread to the interrupt handler, the 2nd interrupt handler for the same IPC interrupt will not happen, and the interrupt thread will not be scheduled more than one time with wrongly.

What it actually fixes:
I hit issue that sometimes the interrupt handler will keep invoking and the irq thread can never be run and then the whole system is hang, on BSW.

Similar issues with other platforms?
I believe we need the similar fixes for BDW/HSW as we don't have a GIE to be disabled there, but for cAVS platforms, we don't need this fix as we disable GIE in the interrupt handler the first time.

File: sound/soc/sof/intel/byt.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/intel/byt.c:248]: @keyonjie I get the rationale for masking the IPC's before waking the thread but you did not answer @plbossart 's question about why you are changing this. 

In my opinion, this check for imrx & SHIM_IMRX_BUSY or imrx & SHIM_IMRX_DONE in the thread is meaningless. You would never get to the thread if they werent true with your changes.

**Comment by plbossart** [comment]
> > > @plbossart @ranj063 Can you consider taking this one please, the logs in #1492 shows the risk without this and test shows no regression with the PR.
> > 
> > 
> > no. I don't have a full picture of why this is needed and what your fix accomplishes and what it actually fixes, and if we have similar issues with other platforms.
> 
> Why this is needed:
> With today's code, on BYT, byt_irq_handler() is entered multiple times for a single IPC interrupt, as we don't mask the interrupt in time in the handler, the next interrupt handler entry(caused by the same IPC interrupt) could happen before the previous interrupt thread mask the interrupt and preempt it. The test code and test result demonstrate this.
> Not talking about the consequence and risk of it, this "byt_irq_handler() is entered multiple times for a single IPC interrupt" goes against our design original intention, isn't it?
> 
> What my fix accomplishes:
> With moving the masking of the SHIM_IMRX_BUSY/DONE bits from the interrupt thread to the interrupt handler, the 2nd interrupt handler for the same IPC interrupt will not happen, and the interrupt thread will not be scheduled more than one time with wrongly.
> 
> What it actually fixes:
> I hit issue that sometimes the interrupt handler will keep invoking and the irq thread can never be run and then the whole system is hang, on BSW.
> 
> Similar issues with other platforms?
> I believe we need the similar fixes for BDW/HSW as we don't have a GIE to be disabled there, but for cAVS platforms, we don't need this fix as we disable GIE in the interrupt handler the first time.

ok, let me make my point very simple:

please work on D0i3 support as your first priority.
When it's complete, we can go back to IPC - and you will have to request QA support.

Multi-tasking across completely unrelated platforms is inefficient. I will not even look at baytrail IPC until D0i3 is complete, merged and fully tested.

**Comment by keyonjie** [comment]
> 
> ok, let me make my point very simple:
> 
> please work on D0i3 support as your first priority.
> When it's complete, we can go back to IPC - and you will have to request QA support.
> 
> Multi-tasking across completely unrelated platforms is inefficient. I will not even look at baytrail IPC until D0i3 is complete, merged and fully tested.

That's true, thanks for your clear point.

**Comment by plbossart** [comment]
replaced by PR #2138

**Pull request closed without merging**